### PR TITLE
Mannequin Packets, Logic + Unlock Quest

### PR DESCRIPTION
**Pull request by zach2good**
image

For the fourth or fifth time in recent memory:
```
That looks easy, shouldn't take me more than a few hours...
                       -zach2good
```

This was a neverending nightmare.

**Review Guide**
- SmallPacket0x03B: Mannequin Equip
- CInventoryCountTo80Packet: Regular and Mannequin update
- Quest: Its_Raining_Mannequins.lua (Not tested yet)
- Fyi_Chalmwoh.lua logic, ie. the mannequin shop/trader
- New bindings to modify exdata of items given to the player: https://github.com/LandSandBoat/server/pull/429/commits/1346b65fbc0d099cc4043297944782f6ce67cfbf

**After noticing that the mannequin exdata is mapped out by Windower after I did Soultrappers that way, I thought this should be quick and easy to figure out**
https://github.com/Windower/Lua/blob/dev/addons/libs/extdata.lua#L1891-L1916

**Inspecting exdata**
image

**Getting UI to respond**

image

**Pestering Windower Folks for clues**

image

**Writing addons to debug**

image

**Packet deep dive**
image

**@InoUno extending PVLV to help me out**
image

**Not pictured: Me running Garrison with Lusi to get my own mannequin so I could test properly!**

**Success!**
image

**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

### CODE REVIEW

**Comment by zach2good** [comment]
Accidentally hit regular submit instead of draft submit again, sorry!

**TODO**
- Test the quest [DONE]
- Maybe implement the posing quest, posing is just another slot in the exdata :D [Nah, not right now]
- More edge cases for mannequins [DONE]
  - Moving weapon from mainhand to offhand, and back, and unequip [DONE]
  - Making sure items are correctly locked [DONE]
  - Moving the mannequin once it's equipped [DONE]
  - Addressing TODO's left in code [DONE]
  - Mannequins aren't updated on first zone-in to MogHouse [DONE]
  - Removing the Mannequin furniture doesn't de-quip everything [DONE]
- Test mannequin NPC a bunch more [DONE]

File: scripts/quests/otherAreas/Its_Raining_Mannequins.lua [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+            onEventFinish =
+            {
+                [311] = function(player, csid, option, npc)
+                    local race = player:getRace()
+                    local chosenMannequin = xi.items.HUME_M_MANNEQUIN + race - 1
+                    if player:getFreeSlotsCount() > 0 and not player:hasItem(chosenMannequin) then
+                        if quest:complete(player) then
+                            player:tradeComplete()
+                            player:addItem({ id = chosenMannequin, exdata = { [18] = race } })
```
Comment (zach2good) [inline scripts/quests/otherAreas/Its_Raining_Mannequins.lua]: Your mannequin will not be useable without setting the race in the exdata!

File: src/map/packets/char.cpp [hunk 2]
```diff
@@ -32,7 +32,7 @@
 CCharPacket::CCharPacket(CCharEntity* PChar, ENTITYUPDATE type, uint8 updatemask)
 {
     this->type = 0x0D;
-    this->size = 0x3A;
+    this->size = 0x5C;
```
Comment (InoUno) [inline src/map/packets/char.cpp]: I'd be careful with this kind of static increase, since it can cause trouble when the server is sending packets to clients (especially if it's a frequent packet). Ideally the size should be variable depending on when the increase is actually needed.

When a packet takes up more space, it means fewer other packets can fit into the same batch of packets sent to clients each time they are sent, since each batch of packets is limited by size. This can cause delays in the queue of packets being sent to clients, which is not good and can lead to client lag.

File: src/map/packets/inventory_count_to_80.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#ifndef _CINVENTORYCOUNTTO80PACKET_H
+#define _CINVENTORYCOUNTTO80PACKET_H
+
+#include "../../common/cbasetypes.h"
+
+#include "basic.h"
+
+class CCharEntity;
+
+class CInventoryCountTo80Packet : public CBasicPacket
```
Comment (InoUno) [inline src/map/packets/inventory_count_to_80.h]: I think we should find a better name for this packet, since this name just seems like a guess at its functionality from PVLV. I don't have a good suggestion at the moment though, and I haven't looked into the other use for the packet (the "count to 80" part).

File: src/map/packets/char.cpp (hunk 2 above)
Comment (zach2good) [reply src/map/packets/char.cpp]: I was honestly just cargo-culting and copying the sizes I observed in captures. It's probably worth trying to cut down all packets to be as small as we can get away with

Comment (TeoTwawki) [reply src/map/packets/char.cpp]: I suggest an overload that does the version you need - some retail packets have multiple alternates in such a manner 
https://github.com/LandSandBoat/server/blob/base/src/map/packets/treasure_lot_item.cpp#L40

though my example is not a great one because the one I linked is now outdated (retail changed what it does, but the old versions still work - lucky for us)

File: src/map/packets/inventory_count_to_80.h (hunk 3 above)
Comment (TeoTwawki) [reply src/map/packets/inventory_count_to_80.h]: maybe just drop the To80 for CInventoryCount

File: scripts/zones/Mhaura/npcs/Fyi_Chalmwoh.lua [hunk 4]
```diff
@@ -3,19 +3,65 @@
...
+        local alreadyOwnedMask = 0
+        player:startEvent(318, 0, alreadyOwnedMask, 100000, 2000, 0, 0, 0, 0)
+    end
 end
 
 entity.onEventUpdate = function(player, csid, option)
 end
 
 entity.onEventFinish = function(player, csid, option)
+    if csid == 318 or csid == 319 then
```
Comment (claywar) [inline scripts/zones/Mhaura/npcs/Fyi_Chalmwoh.lua]: Was just thinking about this block.  Since the items are linear, this could be simplified without even the need for an additional table (We would probably want to sanity check that option is in 1..8, but after that it could be reduced to one line:
```
player:addItem({ id = option + 255, exdata = { [18] = option } })
```

Comment (TeoTwawki) [reply scripts/zones/Mhaura/npcs/Fyi_Chalmwoh.lua]: > Was just thinking about this block. Since the items are linear, this could be simplified without even the need for an additional table (We would probably want to sanity check that option is in 1..8, but after that it could be reduced to one line:
> 
> ```
> player:addItem({ id = option + 255, exdata = { [18] = option } })
> ```

Nope, uncle Teo lesson time: NEVER trust the optioncode in either event finish or event update to actually be legit. they can be faked, so you have to re-validate the items against said code every time.

Comment (TeoTwawki) [reply scripts/zones/Mhaura/npcs/Fyi_Chalmwoh.lua]: I dunno why my brain decided to be concerned there Clay, the existing code still woudl have that problem AND who the F is cheating at mannequin races anyway. I'm gonna blame my lack of sleep.

Comment (claywar) [reply scripts/zones/Mhaura/npcs/Fyi_Chalmwoh.lua]: If that's the case, both versions might have to be rethought.  Do we have a way of knowing which option was selected in onTrigger though?

Comment (TeoTwawki) [reply scripts/zones/Mhaura/npcs/Fyi_Chalmwoh.lua]: damn son yer too quick, was about to hide my posts

> Do we have a way of knowing which option was selected in onTrigger though?

historically we've been using get/setLocalVar to record and read back what the original input was before shenanigans happen.

Comment (TeoTwawki) [reply scripts/zones/Mhaura/npcs/Fyi_Chalmwoh.lua:67]: > Do we have a way of knowing which option was selected in onTrigger though?

historically we've been using get/setLocalVar to record and read back what the original input was before shenanigans happen.

**Comment by zach2good** [comment]
Squashed my commits, tested all my TODOs, opted to only do the first quest in the series, leaving my partial code and notes in place for whoever comes along to do the next quests.

Once this is merged, I want to leave behind an issue for extracting the shared logic out of `packet_system` into `CItemFurnishing`, or something.

Ready for re-review 👍

**Review by claywar — APPROVED** [review]
